### PR TITLE
Service use cohort 

### PR DESCRIPTION
**Pull request by shintoLamp**
Set of functions to create the service use cohort lookup. 

Tested against actual data and I'm happy with the results, but the logic of the helper functions probably needs to be double-checked against the SPSS syntax.

### CODE REVIEW

File: R/create_service_costs.R [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#' Create seperate costs for each service type
+#'
+#' @param data A data frame
+#'
+#' @return A data frame with 14 additional variables, 13 costs and one flag for operations
+#' @export
```
Comment (Jennit07) [inline R/create_service_costs.R:6]: Remove export for non top level functions

Comment (shintoLamp) [reply R/create_service_costs.R]: Done, all functions in main lookup script

File: R/create_service_use_lookup.R [hunk 2]
```diff
@@ -0,0 +1,330 @@
+#' Title Create the Service Use Cohort lookup file
```
Comment (Moohan) [inline R/create_service_use_lookup.R:1]:
```suggestion
#' Create the Service Use Cohort lookup file
```

File: R/create_service_use_lookup.R [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+#'
+#' @param data A data frame
+#' @param year The financial year
+#' @param write_to_disk Default [TRUE]
```
Comment (Moohan) [inline R/create_service_use_lookup.R:5]: Square brackets are for links 😉
```suggestion
#' @param write_to_disk Default `TRUE`
```

File: R/create_service_use_lookup.R [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+#'
+#' @return Either save out a copy of the lookup, or return it for checks
```
Comment (Moohan) [inline R/create_service_use_lookup.R:7]: The `@return` value is what is returned by the function. So in this case that should always be the data. Whether the data gets written to the disk or not is a 'side effect' of the function, which should be documented in the `write_to_disk` param.

File: R/create_service_use_lookup.R (hunk 2 above)
Comment (shintoLamp) [reply R/create_service_use_lookup.R:1]: Done

File: R/create_service_use_lookup.R (hunk 3 above)
Comment (shintoLamp) [reply R/create_service_use_lookup.R:5]: Done, thanks!

File: R/create_service_use_lookup.R (hunk 4 above)
Comment (shintoLamp) [reply R/create_service_use_lookup.R:7]: Cool, thanks for the tip

File: R/create_service_use_lookup.R [hunk 5]
```diff
@@ -0,0 +1,775 @@
...
+
+#' Add operation flag
+#'
+#' @param op1a A vector of operation codes
+#'
+#' @return A boolean vector showing whether a record contains an operation or not
+#' @family Demographic and Service Use Cohort functions
+add_operation_flag <- function(op1a) {
+  operation_flag <- !is_missing(op1a)
+}
```
Comment (Jennit07) [inline R/create_service_use_lookup.R:381]:
```suggestion
return(operation_flag)
```

File: R/create_service_use_lookup.R [hunk 6]
```diff
@@ -0,0 +1,775 @@
...
+#' @param year The financial year
+#' @param write_to_disk Default `TRUE`, will write the lookup to the
+#' Cohorts folder defined by [get_slf_dir]
+#'
+#' @return The Service Use lookup file
+#' @export
+#'
+#' @family Demographic and Service Use Cohort functions
+create_service_use_lookup <- function(data, year, write_to_disk = TRUE) {
+  return_data <- data %>%
```
Comment (Jennit07) [inline R/create_service_use_lookup.R:13]: Would it be useful to have the function `check_variables_exist` here? similar to the demographic cohort code

File: R/create_service_use_lookup.R [hunk 7]
```diff
@@ -0,0 +1,775 @@
...
+#' @param data A data frame
+#' @param year The financial year
+#' @param write_to_disk Default `TRUE`, will write the lookup to the
+#' Cohorts folder defined by [get_slf_dir]
+#'
+#' @return The Service Use lookup file
+#' @export
+#'
+#' @family Demographic and Service Use Cohort functions
+create_service_use_lookup <- function(data, year, write_to_disk = TRUE) {
```
Comment (Jennit07) [inline R/create_service_use_lookup.R:12]:
```suggestion
create_service_use_cohort <- function(data, year, write_to_disk = TRUE) {
```
One small name change suggestion as i said in the demographic cohort script - we may get confused with the social care demographic lookup. Suggesting this one for consistency

File: R/create_service_use_lookup.R [hunk 8]
```diff
@@ -0,0 +1,775 @@
+#' Create the Service Use Cohort lookup file
+#'
+#' @param data A data frame
+#' @param year The financial year
+#' @param write_to_disk Default `TRUE`, will write the lookup to the
+#' Cohorts folder defined by [get_slf_dir]
+#'
+#' @return The Service Use lookup file
```
Comment (Jennit07) [inline R/create_service_use_lookup.R:8]:
```suggestion
#' @return The Service Use cohort file
```

**Comment by Jennit07** [comment]
I've updated the build notes so that they are passing but there is still one test which is failing and i'm not sure why

File: R/create_service_use_lookup.R [hunk 9]
```diff
@@ -0,0 +1,776 @@
...
+      "ae2_cost"
+    )
+
+  if (write_to_disk == TRUE) {
+    write_rds(return_data,
+      path = glue::glue("{get_slf_dir()}/Cohorts/Service_Use_Cohorts_{year}.rds")
+    )
+  } else {
+    return(return_data)
+  }
```
Comment (Jennit07) [inline R/create_service_use_lookup.R:147]: can use the function `get_service_use_cohorts_path` here

File: R/create_service_use_lookup.R (hunk 5 above)
Comment (Moohan) [reply R/create_service_use_lookup.R:382]:
```suggestion
return(operation_flag)
}
```

File: R/create_service_use_lookup.R (hunk 7 above)
Comment (Moohan) [reply R/create_service_use_lookup.R:11]:
```suggestion
create_service_use_cohorts <- function(data, year, write_to_disk = TRUE) {
```

File: R/create_service_use_lookup.R (hunk 9 above)
Comment (Moohan) [reply R/create_service_use_lookup.R:147]:
```suggestion
  if (write_to_disk == TRUE) {
    write_rds(return_data,
      path = get_service_use_cohorts_path(year)
    )
  } 
  
return(return_data)
```

File: R/create_service_use_lookup.R [hunk 10]
```diff
@@ -0,0 +1,776 @@
+#' Create the Service Use Cohort lookup file
```
Comment (Moohan) [inline R/create_service_use_lookup.R:1]:
```suggestion
#' Create the Service Use Cohorts lookup file
```

File: tests/testthat/test-create_service_use_lookup.R [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+      "Residential Care", "Unassigned", "Unassigned", "Unassigned"
+    )
+  )
+})
+
+# test_that("Errors handled as expected", {
+#   expect_error(create_service_instances(tibble::tribble(~cij_marker, ~test_bad)),
+#     regexp = "Variables .+ are required, but are missing from `data`"
+#   )
+# })
```
Comment (Moohan) [inline tests/testthat/test-create_service_use_lookup.R:451]:
```suggestion

```

File: R/create_service_use_lookup.R [hunk 12]
```diff
@@ -0,0 +1,777 @@
...
+#' @param cost_total_net A vector of total net costs
+#'
+#' @return A vector of home care costs
+#' @family Demographic and Service Use Cohort functions
+calculate_home_care_cost <- function(recid, cost_total_net) {
+  home_care_cost <- dplyr::if_else(
+    recid %in% c("HC-", "HC + ", "INS", "RSP", "MLS", "DC", "CPL"), cost_total_net, 0
+  )
+  return(home_care_cost)
+}
```
Comment (Moohan) [inline R/create_service_use_lookup.R:273]: This isn't properly implemented yet so should just return 0
```suggestion
calculate_home_care_cost <- function(recid, cost_total_net) {
  #home_care_cost <- dplyr::if_else(
  # recid %in% c("HC-", "HC + ", "INS", "RSP", "MLS", "DC", "CPL"), cost_total_net, 0
  #)
  
  home_care_cost <- 0
  return(home_care_cost)
}
```

File: R/create_service_use_lookup.R [hunk 13]
```diff
@@ -0,0 +1,777 @@
...
+#' Calculate cost for Care Home records
+#' @description A record is considered to have a home care cost if the recid is CH
+#'
+#' @param recid A vector of record IDs
+#' @param cost_total_net A vector of total net costs
+#'
+#' @return A vector of care home costs
+#' @family Demographic and Service Use Cohort functions
+calculate_care_home_cost <- function(recid, cost_total_net) {
+  care_home_cost <- dplyr::if_else(recid %in% c("CH"), cost_total_net, 0)
```
Comment (Moohan) [inline R/create_service_use_lookup.R:285]:
```suggestion
  care_home_cost <- dplyr::if_else(recid == "CH", cost_total_net, 0)
```

File: R/create_service_use_lookup.R [hunk 14]
```diff
@@ -0,0 +1,777 @@
...
+#' the recid is 01B, the CIJ patient type is Non-Elective, and the specialty is not AB
+#'
+#' @inheritParams calculate_geriatric_cost
+#' @param cij_pattype A vector of CIJ patient types
+#'
+#' @return A vector of Acute Emergency costs
+#' @family Demographic and Service Use Cohort functions
+calculate_acute_emergency_cost <- function(recid, cij_pattype, spec, cost_total_net) {
+  acute_emergency_cost <- dplyr::if_else(recid %in% c("01B") & cij_pattype %in% c("Non-Elective") &
+    !(spec %in% c("AB")), cost_total_net, 0)
```
Comment (Moohan) [inline R/create_service_use_lookup.R:230]:
```suggestion
  acute_emergency_cost <- dplyr::if_else(recid == "01B" & cij_pattype == "Non-Elective" &
    spec != "AB", cost_total_net, 0)
```

File: R/create_service_use_lookup.R [hunk 15]
```diff
@@ -0,0 +1,777 @@
...
+#' @param recid A vector of record IDs
+#' @param cij_pattype A vector of CIJ patient types
+#' @param cost_total_net A vector of total net costs
+#'
+#' @return A vector of Maternity costs
+#' @family Demographic and Service Use Cohort functions
+calculate_maternity_cost <- function(recid, cij_pattype, cost_total_net) {
+  maternity_cost <- dplyr::if_else(
+    recid %in% c("02B") | cij_pattype %in% c("Maternity"), cost_total_net, 0
+  )
```
Comment (Moohan) [inline R/create_service_use_lookup.R:182]:
```suggestion
  maternity_cost <- dplyr::if_else(
    recid == "02B" | cij_pattype == "Maternity", cost_total_net, 0
  )
```

File: R/create_service_use_lookup.R [hunk 16]
```diff
@@ -0,0 +1,777 @@
...
+#' @inheritParams calculate_geriatric_cost
+#' @param cij_pattype A vector of CIJ patient types
+#' @param cij_ipdc A vector of CIJ inpatient/daycase markers
+#'
+#' @return A vector of Acute Elective costs
+#' @family Demographic and Service Use Cohort functions
+calculate_acute_elective_cost <- function(recid, cij_pattype, cij_ipdc, spec, cost_total_net) {
+  acute_elective_cost <- dplyr::if_else(recid %in% c("01B") &
+    (cij_pattype %in% c("Elective") | cij_ipdc %in% c("D")) &
+    !(spec %in% c("AB")), cost_total_net, 0)
```
Comment (Moohan) [inline R/create_service_use_lookup.R:215]:
```suggestion
  acute_elective_cost <- dplyr::if_else(recid == "01B" &
    (cij_pattype == "Elective" | cij_ipdc == "D") &
    spec != "AB", cost_total_net, 0)
```

File: R/create_service_use_lookup.R [hunk 17]
```diff
@@ -0,0 +1,777 @@
...
+#'
+#' @param recid A vector of record IDs
+#' @param cij_pattype A vector of CIJ patient types
+#' @param cost_total_net A vector of total net costs
+#'
+#' @return A vector of Hospital Elective costs
+#' @family Demographic and Service Use Cohort functions
+calculate_hospital_elective_cost <- function(recid, cij_pattype, cost_total_net) {
+  hospital_elective_cost <- dplyr::if_else(
+    recid %in% c("01B", "04B", "50B", "GLS") & cij_pattype %in% c("Elective"), cost_total_net, 0
```
Comment (Moohan) [inline R/create_service_use_lookup.R:301]:
```suggestion
    recid %in% c("01B", "04B", "50B", "GLS") & cij_pattype == "Elective", cost_total_net, 0
```

File: R/create_service_use_lookup.R [hunk 18]
```diff
@@ -0,0 +1,777 @@
...
+#' @description A record is considered to have a Hospital Emergency cost if the recid is
+#' in c("01B", "04B", "50B", "GLS") and the CIJ patient type is Non-Elective
+#'
+#' @inheritParams calculate_hospital_elective_cost
+#'
+#' @return A vector of Hospital Emergency costs
+#' @family Demographic and Service Use Cohort functions
+calculate_hospital_emergency_cost <- function(recid, cij_pattype, cost_total_net) {
+  hospital_emergency_cost <- dplyr::if_else(
+    recid %in% c("01B", "04B", "50B", "GLS") & cij_pattype %in% c("Non-Elective"), cost_total_net, 0
```
Comment (Moohan) [inline R/create_service_use_lookup.R:316]:
```suggestion
    recid %in% c("01B", "04B", "50B", "GLS") & cij_pattype == "Non-Elective", cost_total_net, 0
```

File: R/create_service_use_lookup.R [hunk 19]
```diff
@@ -0,0 +1,777 @@
...
+
+#' Calculate cost for Community Health records
+#' @description A record is considered to have a Community Health cost if the recid is DN
+#'
+#' @inheritParams calculate_prescribing_cost
+#'
+#' @return A vector of Community Health costs
+#' @family Demographic and Service Use Cohort functions
+calculate_community_health_cost <- function(recid, cost_total_net) {
+  community_health_cost <- dplyr::if_else(recid %in% c("DN"), cost_total_net, 0)
```
Comment (Moohan) [inline R/create_service_use_lookup.R:355]:
```suggestion
  community_health_cost <- dplyr::if_else(recid == "DN", cost_total_net, 0)
```

**Review by Moohan — DISMISSED** [review]
Lots of code simplification changes which I can just commit.

```r
bench::mark(
  equal = c("test", "not_test") == "test",
  in_operator = c("test", "not_test") %in% "test",
  in_vector = c("test", "not_test") %in% c("test")
) %>% 
  ggplot2::autoplot()
```

image

**Comment by Moohan** [comment]
Add test at the top to ensure `recid` isn't `NA` with `anyNA()`

**Comment by Moohan** [comment]
There's a conflict being caused because you have a  function called `assign_maternity cohort` on here and for the demog cohorts.

I think we probably want to rename them all e.g. `assign_su_cohort_maternity` etc.

File: R/create_service_use_lookup.R (hunk 6 above)
Comment (Moohan) [reply R/create_service_use_lookup.R:12]: I'll do this

**Comment by Jennit07** [comment]
@Moohan Rename of cohort functions done. Also made changes to demographic cohort functions for consistency. PR open #464

**Comment by Jennit07** [comment]
Tested this using 1920 data and think it's looking good. Ready to merge

**Review by Jennit07 — APPROVED** [review]
Ready to merge

**Pull request merged**